### ISSUE TITLE
Create a testset for manual mapping to ALTO files/pages in original document

### ISSUE THREAD

**Issue by MansMeg** [welfare-state-analytics/riksdagen-corpus#186]
As is being discussed in #179, there seems to be a problem with the mapping between the actual page numbers of the protocols.

We should create a testset for manually assessing the quality of the mapping back to the original pages. i suggest we create a test set in the following way.

- [x] Take a random sample of 3 page XML elements per decade in the data and manually if the page reference is correct.
- [x] If this is done really quickly , add one more random XML page element per decade. I guess roughly 1-2h is enough for this testset.
- [ ] Include the manually curated test set in the corpus as other manually curated data is done. Make it explicit that this is a testset.
- [ ] Estimate the total error based on the corpus data
- [ ] Include the testset in the unit tests - the error rate should not go down, the test should fail.

**Comment by rbbby** [comment]
I finished a sample of 5 pages per decade and all page links were correct. This is very strange considering that almost all page links in #179 were incorrect. Need to investigate what is going wrong here further.

**Comment by MansMeg** [comment]
Ok so already 50 samples? That was quick. Maybe add to 20 pages per decade to be more sure? I guess it is not a huge problem (then we would have seen it) but with 20 per decade it would also give a good estimate.

Have you stored it so we @ninpnin can add it to the corpus?

**Comment by rbbby** [comment]
We only have pages for the OCR:d (not born digital) protocols so it was only 35 pages. I doubled the amount, annotating a total of 70 pages, storing it in _input/segmentation/image_url_sample.csv_.

One error was found in:

> corpus/protocols/198990/prot-198990--61.xml
> https://betalab.kb.se/prot-198990--61/prot_198990__61-122.jp2/_view

Following the pages at betalab and the corpus to see where page numbering / links get out of sync. Starting from page 026 (real page number is 27) with both sources being in sync:

> page,betalab,corpus
027,blank,missing
028,correct,correct
029,missing,blank
030,blank,skips to first content page of bilaga (p3)
031,missing,missing
032,front page of bilaga (p1), (p4)
033,(p2),(p5)

It is very unclear what is going on here with missing or empty pages in either source. The corpus also skips some pages.

**Comment by MansMeg** [comment]
Boom! Very good initiative to look at this @rbbby . Maybe annotate even more (and also born digital) and see if we fin any patterns.

A rough estimate is that 1.4% of the corpus has problems here.

**Comment by ninpnin** [comment]
Seems like the scripts assumes all consecutive pages exist in the KB package, which turns out not to be true in some of the cases. I can write a fix once I can reach KB's servers again.

**Comment by ninpnin** [comment]
As a bonus, we'll get info on what pages are missing on KB's side now :)

**Comment by ninpnin** [comment]
I've implemented a prototype of this. As we do some small preprocessing right away, the test can't directly compare word error rates. Instead, it

1. Extracts all text both from the ALTO file and parlaclarin file
2. Removes whitespace and hyphen characters from both
3. Heuristically splits the text by '.' into sentences
4. Calculates how many of these 'sentences' are exactly the same

A 5% error upper limit is applied. On the protocol I tested this on, there were no discrepancies, i.e. a 0% error rate.

The question is, how many protocols do we want to check everytime we make a push? We can't really test all, as that would be potentially upwards of a hundred gigabytes of data.
